Added backend and Trainer

franchuterivera opened the pull request:
This PR enables to train a pipeline and adds a backend object to it, mainly for logging. 

We will need this backend to store the files for ensemble, but this can be done better when we are coding this later item.

ravinkohli commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,328 @@
+                    )
+            else:
+                string += "\t{}: {}\n".format(
+                    key if 'time' in key else key,
```
maybe a typo?

ravinkohli reviewed (changes requested):
So the code here looks fine. Is the tensorboard_log [here](https://github.com/franchuterivera/Auto-PyTorch_refactor/blob/e6e0dde9a6c766193c859fb64d11ec247715fd53/autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py#L390) left blank for later? Also, is there a way to get the trained model? I think it could be useful to print the pytorch model and return it after training.

franchuterivera commented:
Hi @ravinkohli I have addressed your concerns in the PR. I do see travis examples failing due to a missing file, I am doing a push to fix this too. I will text you once it is done

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        It handles epoch based criteria as well as training based criteria.
+
+
+        # Maybe it is better to use pynisher? Remember the whole everything gets
```
I would prefer to use pynisher, Matthias and I have ownership of that now

franchuterivera replied:
Understood. I have added pynisher support both for memory and for runtime

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+
+        # Maybe it is better to use pynisher? Remember the whole everything gets
+        copied when forking memory, so it makes sense to ask... but also,
+        how to we regulate memory from GPU and CPU?
```
I dont think it is necessary to regulate GPU/CPU memory or do you have a case where it would be? Typically you use all available cores / GPUs and those should be given by the Optimizer/Scheduler

franchuterivera replied:
Yeah, I agree the scheduler should manage. This was just a reminder that the scheduler has to be gpu aware, because we don't want to run out of memory in the GPU because we over scheduled. This was a reminder that pynisher cannot control this today nor smac is aware of this (yet).

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        """
+        self.start_time = time.time()
+        self.budget_type = budget_type
+        self.max_value = max_value
```
Maybe it is better to have a max_epoch and max_time, both with large defaults? Then we could also consider combined limits like "either 50 epochs or 1h runtime"

franchuterivera replied:
Added support for this! Following is now possible:

- epoch only support

- runtime only support

- the first budget that is exhausted be it epoch/time

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        return False
+
+
+class RunSummary(object):
```
Maybe add a description that this is done epoch wise

franchuterivera replied:
I have added more documentation to this class

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        Utility to register global statistic about the model or training,
+        that are agnostic to epochs
+        """
+        self.total_parameter_count = total_parameter_count
```
Is this redundant with the init?

franchuterivera replied:
Yes! I have removed  this

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        model: torch.nn.Module,
+        criterion: torch.nn.Module,
+        budget_tracker: BudgetTracker,
+        optimizer: torch.nn.Module,
```
I thought that torch.optim.optimizer wasnt an nn.Module (https://pytorch.org/docs/stable/_modules/torch/optim/optimizer.html#Optimizer) or am I wrong?

franchuterivera replied:
Thanks for catching this, so much for static checking from mypy :D

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        """
+        return False
+
+    def train(self, train_loader: torch.utils.data.DataLoader, epoch: int
```
Maybe a better name would be train_epoch or something indicating that this is one epoch

franchuterivera replied:
thanks for the suggestion I have renamed this procedure!

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        outputs_data = list()
+        targets_data = list()
+
+        for step, (data, targets) in enumerate(train_loader):
```
This could be outsourced to an additional method `make_step()` such that we could train for a certain number of steps instead of epochs

franchuterivera replied:
thanks for the suggestion I have created this function

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+                    epoch * len(train_loader) + step,
+                )
+
+        return loss_sum / N, self.compute_metrics(outputs_data, targets_data)
```
I think we should add an option to not evaluate train metrics for better anytime performance

franchuterivera replied:
Done in the new push!

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/MixUpTrainer.py:
```diff
@@ -0,0 +1,69 @@
+    def get_hyperparameter_search_space(dataset_properties: typing.Optional[typing.Dict] = None
+                                        ) -> ConfigurationSpace:
+        alpha = UniformFloatHyperparameter(
+            "alpha", 0, 1, default_value=0.5)
```
I think the default should be less aggressive, probably 0.2

franchuterivera replied:
Thanks, I have made this change. I have personally not YET seen this giving really good results, but we will evaluate it once the framework is ready

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,351 @@
+        """
+        raise NotImplementedError()
+
+    def criterion_preparation(self, y_a: np.ndarray, y_b: np.ndarray = None, lam: float = 1.0
```
I assume y_a and y_b are used for mixup. Maybe it would be cleaner to just have y here and in the StandardTrainer and have the MixupTrainer override this method

franchuterivera replied:
This is the one comment I cannot address. Mypy does not allow this redefinition as we need to comply with the parent-child implementation. I used to have it as you proposed (because autopytorch originally had it like this) But with mypy, we need to comply with the parent class interface

LMZimmer replied:
Okay, then let's just go with this. This allows us to introduce weights via y_b which is hopefully enough for most cases

LMZimmer reviewed (changes requested):
Nice work! I left some suggestions

franchuterivera commented on autoPyTorch/pipeline/components/preprocessing/tabular_preprocessing/utils.py:
```diff
@@ -21,9 +21,10 @@ def get_tabular_preprocessers(X: Dict[str, Any]) -> Dict[str, List[BaseEstimator
     for key, value in X.items():
         if isinstance(value, dict):
             if 'numerical' or 'categorical' in value.keys():
```
Yes, I have removed this extra line!

franchuterivera commented on autoPyTorch/pipeline/components/setup/network/base_network.py:
```diff
@@ -152,6 +152,25 @@ def to(self, device: Optional[torch.device] = None) -> None:
+        """
+        Performs batched prediction given a loader object
+        """
+        assert self.network is not None
```
This is exclusive because because of mypy, but yes, user should not see it

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+        if dataset_properties is None:
+            dataset_properties = {}
+
+        # Compile a list of legal preprocessors for this problem
```
Legal trainers?

franchuterivera replied:
Thanks!, I have fixed this

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+            max_epochs=X['epochs'] if 'epochs' in X else None,
+        )
+
+        self.choice.prepare(
```
Maybe we could swap `self.choice` for something more descriptive like `self.trainer_choice`

franchuterivera replied:
This we cannot do, because it is the choice framework for dynamically generating the configuration space.

There are choices and there are components. The self.choice is the product of selecting within standard and mixed training

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+
+        self.choice.prepare(
+            model=X['network'],
+            metrics=[m() for m in get_metrics(X['dataset_properties'])],
```
Are all metrics determined by the dataset properties? I would imaging that the user also wants to specify additional metrics, e.g. log_loss, accuracy and balanced accuracy

franchuterivera replied:
Thanks, I think Ravin added support for this, so I missed to integrate this. I will add it now.

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+        self.choice.prepare(
+            model=X['network'],
+            metrics=[m() for m in get_metrics(X['dataset_properties'])],
+            criterion=get_loss_instance(X['dataset_properties']),
```
Same here, the user possibly wants to specify a custom loss

franchuterivera replied:
Thanks, I think Ravin added support for this, so I missed to integrate this. I will add it now.

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+                if 'test_data_loader' in X and X['test_data_loader']:
+                    test_loss, test_metrics = self.choice.evaluate(X['test_data_loader'], epoch)
+
+            # TODO: does it make sense to call all schedulers here?
```
You could move the scheduler call to the (Base)Trainer, then we can account for calling them step-wise

franchuterivera replied:
yes of course

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+            bool: If true, training should be stopped
+        """
+        assert self.run_summary is not None
+        bad_epochs = self.run_summary.get_best_epoch() - self.run_summary.get_last_epoch()
```
Maybe `epochs_since_best` or similar is more fitting

franchuterivera replied:
I will add this in the next push!

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+        if 'early_stopping' in X and X['early_stopping']:
+            return True
+
+        # We need to know if we should reduce the rate based on val loss
```
Not strictly, that could also be done when evaluating every e.g. 5 epochs. It would be nice if we could support that

franchuterivera replied:
I can just code something quick to support this, but I will rather add proper checking and support for this, and hence I created https://github.com/LMZimmer/Auto-PyTorch_refactor/issues/24

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:
```diff
@@ -0,0 +1,479 @@
+        Returns:
+            torch.device: the device in which to compute operations. Cuda/cpu
+        """
+        if not torch.cuda.is_available():
```
Could we add a logging info here that CPU or GPU is chosen?

franchuterivera replied:
I will add this in next push!

LMZimmer commented on autoPyTorch/pipeline/components/training/trainer/base_trainer.py:
```diff
@@ -0,0 +1,381 @@
+
+
+class BaseTrainerComponent(autoPyTorchTrainingComponent):
+
```
A short docstring might be useful here to state that Trainers have to inherit from this class and that `TrainerChoice` will build on top of this

franchuterivera replied:
I will add this in the next push!

LMZimmer reviewed (changes requested):
Thank you for the fixes, I added just a few small remarks 👍
